runtime/runtime_controller.cc

```diff
@@ -46,48 +46,76 @@ RuntimeController::RuntimeController(
       platform_data_(std::move(p_platform_data)),
       isolate_create_callback_(p_isolate_create_callback),
       isolate_shutdown_callback_(p_isolate_shutdown_callback),
-      persistent_isolate_data_(std::move(p_persistent_isolate_data)) {
-  // Create the root isolate as soon as the runtime controller is initialized.
+      persistent_isolate_data_(std::move(p_persistent_isolate_data)),
+      weak_factory_(this) {
+  // Create the root isolate as soon as the runtime controller is initialized,
+  // but not using a synchronous way to avoid blocking the platform thread a
+  // long time as it is waiting while creating `Shell` on that platform thread.
   // It will be run at a later point when the engine provides a run
   // configuration and then runs the isolate.
-  auto strong_root_isolate =
-      DartIsolate::CreateRootIsolate(
-          vm_->GetVMData()->GetSettings(),                //
-          isolate_snapshot_,                              //
-          task_runners_,                                  //
-          std::make_unique<PlatformConfiguration>(this),  //
-          snapshot_delegate_,                             //
-          io_manager_,                                    //
-          unref_queue_,                                   //
-          image_decoder_,                                 //
-          p_advisory_script_uri,                          //
-          p_advisory_script_entrypoint,                   //
-          nullptr,                                        //
-          isolate_create_callback_,                       //
-          isolate_shutdown_callback_                      //
-          )
-          .lock();
-
-  FML_CHECK(strong_root_isolate) << "Could not create root isolate.";
-
-  // The root isolate ivar is weak.
-  root_isolate_ = strong_root_isolate;
-
-  strong_root_isolate->SetReturnCodeCallback([this](uint32_t code) {
-    root_isolate_return_code_ = {true, code};
-  });
-
-  if (auto* platform_configuration = GetPlatformConfigurationIfAvailable()) {
-    tonic::DartState::Scope scope(strong_root_isolate);
-    platform_configuration->DidCreateIsolate();
-    if (!FlushRuntimeStateToIsolate()) {
-      FML_DLOG(ERROR) << "Could not setup initial isolate state.";
-    }
-  } else {
-    FML_DCHECK(false) << "RuntimeController created without window binding.";
-  }
-
-  FML_DCHECK(Dart_CurrentIsolate() == nullptr);
+  create_and_config_root_isolate_ =
+      std::async(std::launch::deferred, [self = weak_factory_.GetWeakPtr()]() {
```

Comment:
Instead of using `std::async` to start a new thread to run this, can we post this task to the UI thread similar to what had been done in https://github.com/flutter/engine/pull/18225? Then the same reasoning for https://github.com/flutter/engine/pull/18225 probably can also be applied here: if all tasks that rely on the root isolate must run on the UI thread, then posting this root isolation task before all those UI thread tasks could guarantee the safety. CC @jason-simmons @gaaclarke

Reply:
Note that I use `std::launch::deferred` here, and `RuntimeController` is only accessed on the UI task runner, so, 'std::async' here would not start a new thread but run on the UI thread when `create_and_config_root_isolate_.get()` is called.

In case of `Engine::Restart`, it creates a new `RuntimeController` and then runs engine immediately to call `GetRootIsolate` without a chance to run any posted tasks. That is why using `std::async` instead of posting task here, `std::async` creates a future and it is produced when `GetRootIsolate` is first-time called. This also guarantees the root isolate is created before anything that rely on it.